Fast Pair Locator Tag sample: describe the firmware update notification feature and add testing steps for it (#21479)
nrfconnect/sdk-nrf: samples/bluetooth/fast_pair/locator_tag/README.rst

It is located in advertising data when Fast Pair advertising is in the discoverable mode, or in the scan response data when it is in the not discoverable mode.

Android notifications about firmware updates
--------------------------------------------

You can receive Android notifications about the new firmware version for your accessory if you add it during the FMDN provisioning process to the `Find My Device app`_ of your Android device.
The subsequent sections highlight the most important details to have this feature work properly.

The default device model for this sample is configured to support the Android intent feature for firmware updates (the **Firmware Type** option) and to use the `nRF Connect Device Manager`_ application as the Android companion application for this accessory firmware.
See the :ref:`fast_pair_locator_tag_google_device_model` section for configuration details.

This sample also supports the firmware version read operation over Bluetooth and the GATT Device Information Service (DIS).
This mechanism is used by the Android device to read the local firmware version of the Fast Pair accessory in the following cases:

> * during the Fast Pair procedure and the FMDN provisioning operation.
> * asynchronously every 24 hours in the FMDN provisioning state.

Reviewer: ```suggestion
* During the Fast Pair procedure and the FMDN provisioning operation.
* Asynchronously every 24 hours in the FMDN provisioning state.
```

Whenever the Android device reads the local firmware version of this Fast Pair accessory, it compares the local version with the version registered in the **Firmware Version** field of the Google Nearby Console.
If the local version is different than the registered version, the Android device generates the firmware update intent that is sent to the `nRF Connect Device Manager`_ application.
Then, the `nRF Connect Device Manager`_ application processes the intent and displays the user notification about the new firmware version available for this accessory.

.. note::
   The example code for handling the Fast Pair firmware update intents in the Android companion application is available in the `nRF Connect Device Manager GitHub repository`_.

>    The code has been added as part of the `nRF Connect Device Manager GitHub PR with support for Fast Pair firmware update intents`_.

Author: The purpose is to provide the demonstration code for the mobile app developers on how to integrate this feature on the mobile app side.

>    The code has been added as part of the `nRF Connect Device Manager GitHub PR with support for Fast Pair firmware update intents`_.

Reviewer 2: Ok, it feels a bit strange at first glance but makes sense if I think a bit longer about it.

.. note::
   You can execute these testing steps in combination with the :ref:`fast_pair_locator_tag_testing_clock_sync` testing steps, as both test variants require you to wait for more than 24 hours.

To test this feature, complete the following steps:

.. tabs::

   .. group-tab:: nRF52 and nRF53 DKs

      1. Go to the :ref:`fast_pair_locator_tag_testing` section and follow the instructions on performing the FMDN provisioning operation.
      #. Observe that **LED 3** is lit, which indicates that the device is provisioned as an FMDN beacon.
      #. Power off the development kit and wait for 24 hours.
      #. Power off your test Android device.
      #. Power on the development kit.
      #. |connect_terminal_specific|
      #. Power on your test Android device and unlock it after the smartphone screen is turned on.
      #. Wait for the Android device to connect to the development kit, read the local firmware version, and disconnect.
      #. Observe that you get the Android notification about the new firmware update on your test Android device.

         .. figure:: /images/bt_fast_pair_locator_tag_android_fw_update_notification.png

>             :scale: 80 %

Reviewer: The image size needs to be reduced. Please follow the same size like the other .pngs in this document.

            :alt: Firmware update notification in the Android notification center

      #. Upgrade the sample firmware to the latest version:

         1. Set the application firmwareversion to ``v99.99.99`` by modifying the following files:

            * The :file:`VERSION` file.
            * The :file:`configuration/prj.conf` file (the :kconfig:option:`CONFIG_BT_FAST_PAIR_FMDN_DULT_FIRMWARE_VERSION_MAJOR`, :kconfig:option:`CONFIG_BT_FAST_PAIR_FMDN_DULT_FIRMWARE_VERSION_MINOR`, and :kconfig:option:`CONFIG_BT_FAST_PAIR_FMDN_DULT_FIRMWARE_VERSION_REVISION` Kconfig option group).

            The new firmware version should match the version specified in the **Firmware Version** field from the :ref:`fast_pair_locator_tag_google_device_model` section.

>          2. Rebuild the sample application to generate the DFU package.
>          3. Follow the instructions from the :ref:`fast_pair_locator_tag_testing_dfu` section to perform the DFU procedure.

Reviewer: ```suggestion
         #. Rebuild the sample application to generate the DFU package.
         #. Follow the instructions from the :ref:`fast_pair_locator_tag_testing_dfu` section to perform the DFU procedure.
```

      #. Wait for the Android device to connect to the development kit, read the local firmware version, and disconnect.
      #. Observe that you no longer get the Android notification about the new firmware update on your test Android device.

   .. group-tab:: nRF54 DKs

      1. Go to the :ref:`fast_pair_locator_tag_testing` section and follow the instructions on performing the FMDN provisioning operation.
      #. Observe that **LED 2** is lit, which indicates that the device is provisioned as an FMDN beacon.
      #. Power off the development kit and wait for 24 hours.
      #. Power off your test Android device.
      #. Power on the development kit.
      #. |connect_terminal_specific|
      #. Power on your test Android device and unlock it after the smartphone screen is turned on.
      #. Wait for the Android device to connect to the development kit, read the local firmware version, and disconnect.
      #. Observe that you get the Android notification about the new firmware update on your test Android device.

         .. figure:: /images/bt_fast_pair_locator_tag_android_fw_update_notification.png

>             :scale: 80 %

Reviewer: Same as above.

            :alt: Firmware update notification in the Android notification center

      #. Upgrade the sample firmware to the latest version:

>          1. Set the application firmwareversion to ``v99.99.99`` by modifying the following files:

Reviewer: ```suggestion
         a. Set the application firmware version to ``v99.99.99`` by modifying the following files:
```
